src/libPMacc/include/nvidia/warp.hpp

```diff
@@ -0,0 +1,89 @@
+/**
+ * Copyright 2015 Rene Widera
+ *
+ * This file is part of libPMacc.
+ *
+ * libPMacc is free software: you can redistribute it and/or modify
+ * it under the terms of either the GNU General Public License or
+ * the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * libPMacc is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License and the GNU Lesser General Public License
+ * for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * and the GNU Lesser General Public License along with libPMacc.
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+
+#include "types.h"
+#include <device_functions.h>
+
+
+namespace PMacc
+{
+namespace nvidia
+{
+
+/** get lane id of a thread within a warp
+ *
+ * id is in range [0,WAPRSIZE-1]
+ * required PTX ISA >=1.3
+ */
+#if (__CUDA_ARCH__ >= 130)
+DINLINE uint32_t getLaneId()
+{
+    uint32_t id;
+    asm("mov.u32 %0, %laneid;" : "=r" (id));
+    return id;
+}
+#endif
+
+/** get id of a warp within a block
+ *
+ * the result is volatile and can changed every time this function is called
+ * required PTX ISA >=1.3
+ */
+#if (__CUDA_ARCH__ >= 130)
+DINLINE uint32_t getWarpId()
+{
+    uint32_t warpId;
+    asm volatile("mov.u32 %0, %warpid;" : "=r" (warpId));
+    return warpId;
+}
+#endif
+
+/** get maximum number of warps
```

Comment:
IMO thats the maximum what the hardware allows.

CUDA PTX Decumentation Cuda 7.0
```
A predefined, read-only special register that returns the maximum number of warp
identifiers.
```